### PR TITLE
Iotaz migration (QScript)

### PR DESCRIPTION
**Pull request by lukaszwawrzyk**
@djspiewak This PR represents my work on migrating `qscript/package.scala` to iotaz with any dependencies.

It is ready for review. My personal concerns with this code are:
* Materializers for instances. They are rather verbose and pretty ugly. They are so repetetive that I even used IntelliJ live templates to generate most of the code.
The issue I opened in iota regarding instances is [here](https://github.com/frees-io/iota/issues/218). It seems like there is a plan to implement something better over next few months. I was also pointed to scalaz-deriving but I already had noticable amount of code rewriten with Materalizers and did not understand a bit of that. There are currently 30 of those, rewriting them to anything that the rewriting person understands shouldn't take more than one working day.
The good thing about those, is that they just work. It even covered some complex cyclic dependency case with shapeless.Lazy.
* Injectable. I removed `Aux` from there as it caused compiler errors (like AssertionError inside compiler). I made `id` implicit as couple of parts of code rely on it. Actually that code previously used `inject` combined with `Inject.reflexiveInstance`. I hope this Aux thing was not important.
More ugly thing I done is that I moved this thing to `foundation` module to `fp` package. Although it is only file move, package still is `quasar.qscript`. I did it to avoid move refactor back and forth. The move was done because of SubInject macro that injects "subcoproduct" into another coproduct can't be in qscript module as it needs to be in a separate compiler run. It depends on Injectable as it creates it. So I put both of those in foundation as it seemed like reasonable place.
So the question is, what to do with this? Is there a better place to put it, a new module perhaps? If it should stay where it is, should I change package of Injectable to be `quasar.fp`?

### CODE REVIEW

**Comment by lukaszwawrzyk** [comment]
I updated iotaz dependency to 0.3.8-SNAPSHOT as it contains my fix for missing hashCode implementation. I merged with master. The build is passing now.
I have further coproducts migrated in `iotaz-migration-2` branch. I will create PR after this one is merged.

I also did some quick compilation performance tests. `qscript` module compiles ~3.2 times faster. The overall compilation time went down from 21.5 to 16 minutes.
Although those "benchmarks" were not anything special, in all the comparisons compilation was consistently faster for the iotaz version.

**Comment by djspiewak** [comment]
I'll take a look at this asap. Nice work! Regarding `Injectable`, I would say just change the package name and leave it where it is. I'm not too fussed about where it sits. The materializer stuff is more annoying, but there's not a lot of good ways around it right now, and using scalaz-deriving would almost certainly give back all the compilation time gains we just achieved (also likely wouldn't work).

**Comment by djspiewak** [comment]
Oh, side-note… I'm really reticent to depend on anything `-SNAPSHOT` due to the special semantics that maven applies to that version quantifier: it makes the build non-reproducible. I would rather publish a hash version snapshot of iotaz so that things are stable and then depend on that. I can take care of that and submit a PR to you.

File: foundation/src/main/scala/quasar/fp/SubInject.scala [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    @SuppressWarnings(Array("org.wartremover.warts.Null"))
+    def apply[Inner[a] <: CopK[_, a], Outer[a] <: CopK[_, a]]: Injectable[Inner, Outer] = macro iotaz.internal.SubInjectMacros.create[Inner, Outer]
+  }
+
+}
+
+// iotaz package was used to get access to packed private code, mostly for extracting
+// type list from CopK. Also most of the code here that is not specific to SubInject
+// was copied from iotaz (because it was private).
+package iotaz.internal {
```
Comment (djspiewak) [inline foundation/src/main/scala/quasar/fp/SubInject.scala]: Can we move this to a separate file?

File: mimir/src/main/scala/quasar/mimir/MapFuncPlanner.scala [hunk 2]
```diff
@@ -38,22 +43,44 @@ object MapFuncPlanner {
...
+    def materialize(offset: Int): MapFuncPlanner[T, G, CopK[LL, ?]]
+  }
 
-        def apply[A <: SourceType](id: cake.trans.TransSpec[A]): AlgebraM[F, Coproduct[G, H, ?], TransSpec[A]] =
-          _.run.fold(G.plan(cake).apply(id), H.plan(cake).apply(id))
-      }
+  object Materializer {
+    @SuppressWarnings(Array("org.wartremover.warts.NonUnitStatements"))
+    implicit def base[T[_[_]], G[_]]: Materializer[T, G, TNilK] = new Materializer[T, G, TNilK] {
+      override def materialize(offset: Int): MapFuncPlanner[T, G, CopK[TNilK, ?]] = ???
```
Comment (djspiewak) [inline mimir/src/main/scala/quasar/mimir/MapFuncPlanner.scala]: Why is this here? If it doesn't exist, we can just remove `base` altogether. If it does exist, then we need to implement the `???`.

File: foundation/src/main/scala/quasar/fp/Injectable.scala [hunk 3]
```diff
@@ -17,7 +17,8 @@
 package quasar.qscript
```
Comment (alissapajer) [inline foundation/src/main/scala/quasar/fp/Injectable.scala]: Re your question in the PR description, let's keep this in the foundation project and put it under package quasar.fp.

**Comment by djspiewak** [comment]
Ok, it looks like iotaz's settings are dictated by sbt-freestyle and sbt-org-plugin, which enforce sonatype without any option of bintray, which in turn means that it's not really possible for me to meaningfully publish a stable hash snapshot. *sigh*

File: effect/src/main/scala/quasar/effect/Capture.scala [hunk 4]
```diff
@@ -55,11 +57,18 @@ sealed abstract class CaptureInstances extends CaptureInstances0 {
...
 
+  implicit def freeCaptureCopK[F[_]: Capture, S[a] <: ACopK[a]](implicit I: F :<<: S): Capture[Free[S, ?]] =
+    new FreeCapture[F, S](I.inj)
+}
+
+sealed abstract class CaptureInstances0 extends CaptureInstances1 {
+
   implicit def freeCapture[F[_]: Capture, S[_]](implicit I: F :<: S): Capture[Free[S, ?]] =
-    new FreeCapture[F, S]
+    new FreeCapture[F, S](injectNT[F, S])
```
Comment (wemrysi) [inline effect/src/main/scala/quasar/effect/Capture.scala]: Alternatively, `new FreeCapture[F, S](I)` now that `Inject` extends `~>`.

File: connector/src/main/scala/quasar/connector/BackendModule.scala [hunk 5]
```diff
@@ -194,17 +196,17 @@ trait BackendModule {
 
   // everything abstract below this line
 
-  type QS[T[_[_]]] <: CoM
```
[the remark this answers is not in the document]
Comment (lukaszwawrzyk) [reply connector/src/main/scala/quasar/connector/BackendModule.scala]: I will verify again on friday, but pretty sure I tried to delete it, but it was used in `fs` related stuff.

[the remark this answers is not in the document]
Comment (lukaszwawrzyk) [reply connector/src/main/scala/quasar/connector/BackendModule.scala]: Yeah, it is used quite a lot. To name one place: quasar.repl.Main#ReplEff. There are 23 usages of `:/:` symbol.

**Comment by lukaszwawrzyk** [comment]
@wemrysi @alissapajer @djspiewak I fixed all the issues.

File: connector/src/main/scala/quasar/connector/BackendModule.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (alissapajer) [reply connector/src/main/scala/quasar/connector/BackendModule.scala]: :+1: 

Seeing the compile time improvements we gained with these changes, I'd definitely like to switch us over to use iotaz coproducts everywhere. Not in this PR though of course :)

[the remark this answers is not in the document]
Comment (lukaszwawrzyk) [reply connector/src/main/scala/quasar/connector/BackendModule.scala]: I was quite far with this refactoring although there were some problems that should mostly go away once quasar.fs is refactored to not use coproducts at all. So the further migration is blocked on that refactoring.

[the remark this answers is not in the document]
Comment (alissapajer) [reply connector/src/main/scala/quasar/connector/BackendModule.scala]: Yeah, we're rewriting the entire quasar file system, so I wouldn't spend time in that part of the code. That will all be ripped out as we're implementing the new file system from scratch. /cc @wemrysi
